### PR TITLE
refactor site_info to use attrs definition

### PR DESCRIPTION
**Pull request by bayc**
# Refactor site_info to use the attrs library and method of class definition

This pull request refactors the `SiteInfo` class to use the attrs library dataclass functionality. This cleans up the `__init__` function and doesn't require the user to write some boilerplate code for each class. It also adds the ability to define converters and validators for input, building robustness into the code.

## Related issue
#145 

## Impacted areas of the software
`site_info.py`

## Additional supporting information
None

## Test results, if applicable
None

### CODE REVIEW

**Review by camirmas — COMMENTED** [review]
Looks good, I left a few comments. I'd also like to discuss a plan for test fixes to go along with this before we merge it

File: hopp/simulation/technologies/sites/site_info.py [hunk 1]
```diff
@@ -103,65 +115,58 @@ def __init__(self, data,
...
-            self.wind_resource = WindResource(data['lat'], data['lon'], data['year'], wind_turbine_hub_ht=hub_height,
-                                            filepath=wind_resource_file)
-            n_timesteps = len(self.wind_resource.data['data']) // 8760 * 8760
-            if self.n_timesteps is None:
-                self.n_timesteps = n_timesteps
-            elif self.n_timesteps != n_timesteps:
-                raise ValueError(f"Wind resource timesteps of {n_timesteps} different than other resource timesteps of {self.n_timesteps}")
+            self.wind_resource = WindResource(self.data['lat'], self.data['lon'], self.data['year'], wind_turbine_hub_ht=self.hub_height,
+                                            filepath=self.wind_resource_file)
+            self.n_timesteps = 8760
```
Comment (dguittet) [inline hopp/simulation/technologies/sites/site_info.py:146]: Why was the calculation of the n_timesteps changed?

File: hopp/simulation/technologies/sites/site_info.py [hunk 2]
```diff
@@ -103,65 +115,58 @@ def __init__(self, data,
...
-        else:
-            logger.info(
-                "Set up SiteInfo with solar and wind resource files: {}, {}".format(self.solar_resource.filename,
-                                                                                    self.wind_resource.filename))
+        # if 'no_wind' in data:
+        #     logger.info("Set up SiteInfo with solar resource files: {}".format(self.solar_resource.filename))
+        # else:
+        #     logger.info(
+        #         "Set up SiteInfo with solar and wind resource files: {}, {}".format(self.solar_resource.filename,
+        #                                                                             self.wind_resource.filename))
```
Comment (dguittet) [inline hopp/simulation/technologies/sites/site_info.py:169]: Why was this commented out?

File: hopp/simulation/technologies/sites/site_info.py [hunk 3]
```diff
@@ -203,71 +208,9 @@ def plot(self,
         axes.set_aspect('equal')
         axes.set(xlim=(min_plot_bound[0], max_plot_bound[0]), ylim=(min_plot_bound[1], max_plot_bound[1]))
         plot_shape(figure, axes, self.polygon, '--', color=border_color, alpha=alpha, linewidth=linewidth / 2)
-        if isinstance(self.polygon, Polygon):
-            shape = [self.polygon]
-        elif isinstance(self.polygon, MultiPolygon):
-            shape = self.polygon.geoms
-        for geom in shape:    
-            xs, ys = geom.exterior.xy    
-            plt.fill(xs, ys, alpha=0.3, fc='g', ec='none')
```
Comment (dguittet) [inline hopp/simulation/technologies/sites/site_info.py:212]: why was this removed?

File: hopp/simulation/technologies/sites/site_info.py [hunk 4]
```diff
@@ -203,71 +208,9 @@ def plot(self,
...
-                except:
-                    valid_region = valid_region.difference(transform(project, make_valid(pm.geometry)))
-        return k, valid_region, lat, lon
-
-    @staticmethod
-    def append_kml_data(kml_data, polygon, name):
-        folder = kml_data._features[0]._features[0]
-        new_pm = kml.Placemark(name=name)
-        new_pm.geometry = polygon
-        folder.append(new_pm)
```
Comment (dguittet) [inline hopp/simulation/technologies/sites/site_info.py:273]: why was this removed? this is needed for reading the site boundary from a KML file

**Comment by camirmas** [comment]
This has fallen behind somewhat, I'm going to close for now and open a new PR with these changes on top of the updated (tests fixed) `dev/refactor`
